internal/xsql/conn.go

```diff
@@ -360,3 +364,311 @@ func (c *conn) BeginTx(ctx context.Context, txOptions driver.TxOptions) (_ drive
+	}
+	return columns, nil
+}
+
+func (c *conn) GetColumnType(ctx context.Context, tableName, columnName string) (dataType string, err error) {
+	var (
+		cn      = c.connector.Connection()
+		session = c.session
+	)
+
+	tableExist, err := helpers.IsTableExists(ctx, cn.Scheme(), tableName)
+	if err != nil {
+		return "", err
+	}
+	if !tableExist {
+		return "", fmt.Errorf("table '%s' not exist", tableName)
+	}
+
+	columnExist, err := c.IsColumnExists(ctx, tableName, columnName)
+	if err != nil {
+		return "", err
+	}
+	if !columnExist {
+		return "", fmt.Errorf("column '%s' not exist in table '%s'", columnName, tableName)
+	}
+
+	err = retry.Retry(ctx, func(ctx context.Context) (err error) {
+		desc, err := session.DescribeTable(ctx, tableName)
+		if err != nil {
+			return err
+		}
+		for _, col := range desc.Columns {
+			if col.Name == columnName {
+				dataType = col.Type.Yql()
+				break
+			}
+		}
+		return nil
+	}, retry.WithIdempotent(true))
+	if err != nil {
+		return "", err
+	}
+	return dataType, nil
+}
+
+func (c *conn) GetPrimaryKeys(ctx context.Context, tableName string) (pkCols []string, err error) {
+	var (
+		cn      = c.connector.Connection()
+		session = c.session
+	)
+
+	tableExist, err := helpers.IsTableExists(ctx, cn.Scheme(), tableName)
+	if err != nil {
+		return nil, err
+	}
+	if !tableExist {
+		return nil, fmt.Errorf("table '%s' not exist", tableName)
+	}
+
+	err = retry.Retry(ctx, func(ctx context.Context) (err error) {
+		desc, err := session.DescribeTable(ctx, tableName)
+		if err != nil {
+			return err
+		}
+		pkCols = append(pkCols, desc.PrimaryKey...)
+		return nil
+	}, retry.WithIdempotent(true))
+	if err != nil {
+		return nil, err
+	}
+	return pkCols, nil
+}
+
+func (c *conn) IsPrimaryKey(ctx context.Context, tableName, columnName string) (ok bool, err error) {
+	cn := c.connector.Connection()
+
+	tableExist, err := helpers.IsTableExists(ctx, cn.Scheme(), tableName)
+	if err != nil {
+		return false, err
+	}
+	if !tableExist {
+		return false, fmt.Errorf("table '%s' not exist", tableName)
+	}
+
+	columnExist, err := c.IsColumnExists(ctx, tableName, columnName)
+	if err != nil {
+		return false, err
+	}
+	if !columnExist {
+		return false, fmt.Errorf("column '%s' not exist in table '%s'", columnName, tableName)
+	}
+
+	pkCols, err := c.GetPrimaryKeys(ctx, tableName)
+	if err != nil {
+		return false, err
+	}
+	for _, pkCol := range pkCols {
+		if pkCol == columnName {
+			ok = true
+			break
+		}
+	}
+	return ok, nil
+}
+
+func (c *conn) GetTables(ctx context.Context, root string) (tables []string, err error) {
+	var (
+		cn           = c.connector.Connection()
+		schemeClient = cn.Scheme()
+	)
+
+	ignoreDirs := map[string]bool{
+		".sys":        true,
+		".sys_health": true,
+	}
+
+	canEnter := func(dir string) bool {
+		if _, ignore := ignoreDirs[dir]; ignore {
+			return false
+		}
+		return true
+	}
+
+	queue := make([]string, 0)
+	queue = append(queue, root)
+
+	for st := 0; st < len(queue); st++ {
```

Comment:
I think no need to iterate over subdirs. This is feature is not allowed in any RDBMs, only YDB as I know. If some app with xorm need to restrict access with some directory - this code must be define table path prefix explicitely.

Reply:
If we will want to get all tables including embedding tables in subdirs - we add code like this

Reply:
At now it is overengenering